prereise/gather/griddata/hifld/data_process/topology.py

```diff
@@ -217,3 +218,100 @@ def get_mst_edges(lines, substations, **kwargs):
         with open(os.path.join(cache_dir, f"mst_{cache_hash}.pkl"), "wb") as f:
             pickle.dump(mst_edges, f)
     return mst_edges
+
+
+def add_interconnects_by_connected_components(
+    lines,
+    substations,
+    seams_substations,
+    substation_assumptions,
+    line_assumptions,
+    interconnect_size_rank,
+):
+    """Disconnect a large connected component using a set of connecting substations,
+    label the resulting connected components, then reconnect dropped lines and
+    substations, using explicit assumptions plus inference from neighboring lines. The
+    ``lines`` and ``substations`` data frames are modified inplace with a new
+    'interconnect' column.
+
+    :param pandas.DataFrame lines: data frame of line information.
+    :param pandas.DataFrame lines: data frame of substations information.
+    :param iterable seams_substations: IDs of substations to drop.
+    :param dict substation_assumptions: labeling assumptions for substations.
+    :param dict line_assumptions: labeling assumptions for lines.
+    :param iterable interconnect_size_rank: ordered iterable of interconnection names.
+    :raises ValueError: if at least one dropped line's interconnection isn't specified
+        and can't be inferred from its neighbors, or if the final number of connected
+        components don't match the length of the ``interconnect_size_rank``.
+    """
+    # Create a graph of the network, and drop lines based on substations
+    dropped_lines = lines.loc[
+        lines["SUB_1_ID"].isin(seams_substations)
+        | lines["SUB_2_ID"].isin(seams_substations)
+    ]
+    g = nx.convert_matrix.from_pandas_edgelist(lines, "SUB_1_ID", "SUB_2_ID")
+    g.remove_nodes_from(seams_substations)
+    # Label interconnections based on their sizes
+    sorted_interconnects = sorted(nx.connected_components(g), key=len)[::-1]
+    labels = pd.Series("unknown", index=lines.index, dtype="string")
+    for i, name in enumerate(interconnect_size_rank):
+        labels.loc[lines.SUB_1_ID.isin(sorted_interconnects[i])] = name
+    labels.loc[dropped_lines.index] = "dropped"
+    # Label some dropped lines and unknown interconnection lines (small islands)
+    for interconnect, sub_ids in substation_assumptions.items():
+        labels.loc[
+            lines["SUB_1_ID"].isin(sub_ids) | lines["SUB_2_ID"].isin(sub_ids),
+        ] = interconnect
+    for interconnect, line_ids in line_assumptions.items():
+        labels.loc[line_ids] = interconnect
+    # Use neighboring lines at non-dropped substations to infer line interconnections
+    dropped_lines = lines.loc[labels == "dropped"]
+    non_dropped_lines = lines.loc[labels != "dropped"]
+    for id, line in dropped_lines.iterrows():
+        non_dropped_sub = (  # noqa: F841
+            line.SUB_1_ID if line.SUB_2_ID in seams_substations else line.SUB_2_ID
```

Comment:
So it is guaranteed there are no lines with both ends in seams substations, right?

Reply:
It is not guaranteed, in fact it does happen in our data set! The path of connection from the Eastern interconnect to the Oklaunion B2B station appears to go from a full-Eastern substation through another intermediate mixed-interconnect substation first before getting to the B2B substation, so both the B2B substation and the 'intermediate' are considered to be 'part of the seam', since both have buses in both interconnects, and the line that connects the two has both ends 'on the seam'. With the input data we have right now, the interconnects still do end up getting separated properly, but there's definitely potential for this sort of edge case to bite us with different input data. This would also preclude your other refactor idea of always looking to the substation at the 'other end' to determine which interconnect a line should connect to within a dropped substation.